### ISSUE TITLE
StereoEnhancer changes the amplitude of the sound

### ISSUE THREAD

**Issue by Umcaruje** [LMMS/lmms#847]
The stereo enhancer alters the volume of the left or the right channel when stereo seperating. This is a pretty big bug, considering a lot of users use stereo widening in their productions.
A test LMMS project: http://lmms.sourceforge.net/lsp/index.php?action=show&file=5552
Listen to the file: http://goo.gl/hQtyl4
How the file looks in audacity: 
![the bug](https://cloud.githubusercontent.com/assets/6282045/3262731/13966190-f269-11e3-805c-dc9d4f8e5e10.PNG)

**Comment by tresf** [comment]
Per @Umcaruje, milestoning and tagging so this gets some attention prior to the release of 1.2.

Not sure how practical this is, but we can bump it if needed.

-Tres

**Comment by Umcaruje** [comment]
So after some talk over at IRC, @curlymorphic and myself agreed that this should be fixed by adding a new plugin that would apply the [Haas Effect](http://en.wikipedia.org/wiki/Precedence_effect). The production of the new plugin shouldn't be too complex to do, since the delay effect has some code that can be used.

The plugin could be called __Haassan__ (That's my idea at least).

**Comment by Sti2nd** [comment]
So that means you aren't fixing anything?

**Comment by Umcaruje** [comment]
Well changing the effect would break backwards compatibility.

**Comment by Sti2nd** [comment]
Hmm, and it could be debated wether this is a bug or not, so ok, makes some sense. You don't feel it will be annoying to always add two plugins just to use the stereo enhancer?

**Comment by Umcaruje** [comment]
? 

The new plugin would be a standalone, no need to add the old one...

**Comment by Sti2nd** [comment]
Ooh, cause that is another way of achieving stereo effect. Ok, so basically a completely new effect that has nothing to do with the old one. Now I get it. I thought you were going to clone the old one and in addition apply the haas effect and then call it "Stereoenhancer New", only that you wanted to call it Haassan.

**Comment by curlymorphic** [comment]
@diizy wrote 
> If we want an improved
> stereo widening effect, it can be done as a new effect plugin. If you
> have an idea for a better implementation of a widening algorithm, feel
> free to code it...

This is why I suggested implementing a new plugin.

**Comment by Sti2nd** [comment]
Cool. Uros used the word fix. Sorry for not understanding, probably the best thing to do. :+1:

**Comment by tresf** [comment]
Yeah it is technically an enhancement then, but symantecs.... :)

We do need to consolidate/hide/renove some of our plugins for 2.0 so perhaps the old version can be part of a cleanup task.  I'm also interested in how much of a change this is in code.  Perhaps we entertain a config option for old behavior and use compat code to turn it on/off.

**Comment by curlymorphic** [comment]
> Perhaps we entertain a config option for old behavior and use compat code to turn it on/off.

I am confused, Looking into the code revels the bug in this reported by  @softrabbit has already been fixed .  

https://github.com/LMMS/lmms/blob/master/include/DspEffectLibrary.h#L336
````
			const float toRad = F_PI / 180;
			const sample_t tmp = inLeft;
			inLeft += inRight * sinf( m_wideCoeff * ( .5 * toRad ) );
			inRight -= tmp * sinf( m_wideCoeff * ( .5 * toRad ) );
````

Currently The stereoEnhancer uses 2 techniques, The Haas Effect, and Polar Exploration.
https://www.soundonsound.com/sos/sep10/articles/pt-0910.htm#Top.

@Umcaruje does this solve your use case, or do you still require a Hass only effect?

**Comment by karmux** [comment]
When I export this project to both ogg and wav then Audacity shows totally different soundwaves.
http://www.upload.ee/image/4617598/lmms-stereo-ogg-vs-wav.png

**Comment by Wallacoloo** [comment]
Could this possibly just be due to how ogg encodes stereo information?

If possible, you should try exporting as wav, and then re-encoding that as
an ogg using an external program (audacity if possible, or maybe ffmpeg).
If the ogg looks similar to that generated directly by lmms, then this is
likely not an issue with lmms.

**Comment by rubiefawn** [comment]
The StereoEnhancer changes the amplitude of my sound IN LMMS, and during export too.

**Comment by karmux** [comment]
@Wallacoloo they also sound differently. When I convert wav to ogg with Audacity then this ogg sounds and looks exactly like wav. It means that LMMS exports differently sounding ogg and wav out of this project. I did some testing about this issue and I couldn't reproduce it with other plugins that I tested. Should be only StereoEnhancer that makes LMMS to export drastically different wav and ogg sounds.

**Comment by curlymorphic** [comment]
> @diizy wrote
> 
> If we want an improved
> stereo widening effect, it can be done as a new effect plugin. If you
> have an idea for a better implementation of a widening algorithm, feel
> free to code it...

This is the reason I dont want to touch the original plugin.

Im unsure how to proceed, do I change the current plugin, or write a seperate one, using the same techiques ( the processes used are pretty standard ) that we can address these issues on?

I dont really like the thought of adding loads of hacky ````if(oldVersion) .... else .....````

@tresf  @diizy @Lukas-W  @ anyoneIForgotPleaseAddThem

what are your thoughs

**Comment by tresf** [comment]
@curlymorphic then please write a new one.  I just want the team to agree that we don't need two of every plugin.  Eventually we need a way of deprecating stuff so we don't end up with a bunch of old unused code. :+1:

**Comment by Wallacoloo** [comment]
The *original* problem in this thread is simply that the wave and ogg
export formats are visibly/audibly different. It probably doesn't have
anything to do with the stereo expansion effect at all, but with whatever
code encodes the ogg.

**Comment by tresf** [comment]
@Wallacoloo yeah, or perhaps a bug in the default export resolution settings of OGG vs WAV per "... not all of the parameters above apply for all formats".

**Comment by curlymorphic** [comment]
@curlymorpic scrolls to top of page to check title,  StereoEnhancer changes the amplitude of the sound

Have some wires been crossed?

**Comment by curlymorphic** [comment]
@Umcaruje are you still up for designing the ui for this?

**Comment by Umcaruje** [comment]
Yes I am :3 Just  tell me how many knobs. 

P.S. is the name gonna be 'Haassan'?

**Comment by curlymorphic** [comment]
> Just tell me how many knobs.

2 knobs please. one for hass delay time, and one for polar exploration amount.


> P.S. is the name gonna be 'Haassan'?

fine by me :)

**Comment by softrabbit** [comment]
Just in case @curlymorphic decides to copy the code for the next generation of the plugin...

This bit seems wrong. Didn't catch it earlier, but now I'm thinking both lines should have a ```-=``` operator. Better late than never, I suppose.
https://github.com/LMMS/lmms/blob/master/include/DspEffectLibrary.h#L338
```
inLeft += inRight * sinf( m_wideCoeff * ( .5 * toRad ) );
inRight -= tmp * sinf( m_wideCoeff * ( .5 * toRad ) );
```

I'm saying both channels should have some of the opposite channel mixed in _with reverse polarity_, not like this (looks more like a butchered MS encoder).

**Comment by curlymorphic** [comment]
I have a prototype demonstrating 3 different algorithms, It is also dual band. 

I would really like some feedback back please., on what algorithms we should use, and if the multiband is a good idea.

Here is a release with a windows32 installer, The plugin to use is called 'Haassan'

https://github.com/curlymorphic/lmms/releases/tag/HassanPrototype1

**Comment by tresf** [comment]
> I would really like some feedback back please., on what algorithms we should use, and if the multiband is a good idea.

Sorry if this is implied, but can you please direct this comment at someone you feel can help?   I say this because the Windows installer won't necessarily attract many developers that know algorithms (are you asking for testers to give feedback too or am I over thinking this?).  The clearer you can be with your feedback, the better feedback I think you'll receive.

**Comment by Umcaruje** [comment]
I'll test it out later today.

**Comment by curlymorphic** [comment]
Sorry I didnt explain properly. I was looking for users to try out the 3 different types of widening, to see If there are any preferences as to what method should be used.

@Umcaruje  @tresf @softrabbit @Sti2nd @unfa @grejppi 

I would really appreciate your views on the audio produced using the 3 different methods provided in the 'Haassan' pluging

https://github.com/curlymorphic/lmms/releases/tag/HassanPrototype1

**Comment by rubiefawn** [comment]
Doesn't Ableton have a stereo widener in Utility? I think it widens sound by using phase cancellation to remove sounds that occur in both left and right.

**Comment by StakeoutPunch** [comment]
That method would result in no sound if you had a completely mono sound source...

**Comment by curlymorphic** [comment]
Out of the 3 types of separation offered, what ones(s) do we want to use?

Do you feel the multiband feature is useful?

Should we have a "led button" to select what effects are active, or should we leave them all running in series, as is the current case? 

There was some feedback given on irc, @Umcaruje  @softrabbit  can you please give your thoughts here.

Once this is determined @Umcaruje can work on the ui.

My votes are to use the hassan, and width effects, and drop the polar. I would also like to keep the multiband feature.

The width effect can be used to narrow a signal, This i personally use injunction with with the multiband feature, to narrow the bottom end of a signal, while leaving the rest to use the stereo field.

I am going to spend some time now, removing the buzzing from the hass control.

**Comment by Umcaruje** [comment]
Changed the milestone, as I don't believe this issue is crucial to 1.2, and I've also just learned that we include the 'Calf Stereo Tools' as a ladspa plugin, which can serve as a replacement.

**Comment by IanCaio** [comment]
I ran into this issue by chance looking for another one. Just adding a comment to link that issue to #5956 . I'm not sure it completely solves the issue but the output gain knob at least makes it more convenient to compensate the undesirable increase in the amplitude.

**Comment by Rossmaxx** [comment]
Since the talk 'was' about adding a new plugin in favour of the current plugin, I'm removing the milestone on this.
